### PR TITLE
Fewer interpolation options

### PR DESCRIPTION
**Pull request by alisterburt**
This PR restricts the interpolation modes available for 2D images as suggested by @tlambert03 in #2546 

The new list of available options was selected to be representative of the whole set (see discussion in #2546) and includes
- nearest
- bilinear
- bicubic
- spline36
- kaiser

### CODE REVIEW

**Comment by sofroniewn** [comment]
Very nice @alisterburt - I do wonder if we want to do some form of deprecation here and a warning, otherwise this will be an immediate breaking change (which we're trying to minimize). We could remove completely from the drop-down and deprecate in the API (I could imagine that we maintain a normal and deprecated Enum and then check if string is in deprecated one for example).

Also it might be nice to add to the doc strings some of the info learned in #2546 about the 5 different types, maybe even link to a wikipedia page or something (not sure how others feel about that!)

**Comment by tlambert03** [comment]
We don't necessarily need to remove programmatic access to those modes, right?  Just reduce the options in the combo box in the gui

**Comment by alisterburt** [comment]
Great suggestions @sofroniewn ! I'll add a deprecated enum and throw some warnings - is there a mechanism in place so that we remember to remove things marked as deprecated?

I like the idea of adding some info in the docstring too - maybe with a link to the matplotlib comparison Talley showed

An alternative is exposing all options via the API and maintaining a separate enum for those available from the dropdown... I didn't go with this because there was the idea of reducing code complexity in the discussion, any thoughts?

Edit: wires crossed @tlambert03 - I agree with you but didn't go that route because there was some discussion about eventually unifying 2d/3d, very happy to redo the PR that way though!

**Comment by tlambert03** [comment]
> because there was some discussion about eventually unifying 2d/3d, very happy to redo the PR that way though!

yeah... that would be nice, but it will take a fair amount of work on the 3D side, and personally, I'm not too worried about it any time soon

**Comment by sofroniewn** [comment]
Given discussion between @tlambert03 and @alisterburt above, my vote would be to just drop the other options from the drop down menu and leave all of them in the API. We already do that for the colormaps and I think is a reasonable approach. I could go either way though if others have strong feelings

> I like the idea of adding some info in the docstring too - maybe with a link to the matplotlib comparison Talley showed

I still like this idea too

**Comment by jni** [comment]
:point_up: :stuck_out_tongue_winking_eye:

**Comment by jni** [comment]
I agree re leaving API working, removing from the UI.

**Comment by tlambert03** [comment]
@alisterburt, this still needs to be modified so-as not to break the API, correct?

**Comment by alisterburt** [comment]
@tlambert03 yep - I have a version with two Enums in place but it doesn't work as I expected. I'll have a look at it now again, will push so you can take a look if you like

edit: there now, setting to an Interpolation not available from the GUI raises a `ValueError` 
```
ValueError: 'hamming' is not a valid Interpolation
```

edit2: there's a very stupid reason for this, fixing now 🤭

**Comment by alisterburt** [comment]
Okay @tlambert03 this now 'works' in that values not in the dropdown can be set programmatically but it blanks out the combobox, would be nicer if that updated properly so will try to fix that now

Edit: works properly - looking at adding tests but might not be done before I go to sleep

File: napari/layers/image/_image_constants.py [hunk 1]
```diff
@@ -29,6 +29,23 @@ class Interpolation(StringEnum):
...
+            Interpolation.BICUBIC,
+            Interpolation.BILINEAR,
+            Interpolation.KAISER,
+            Interpolation.NEAREST,
+            Interpolation.SPLINE36,
+        ]
+    ],
+)
+
+
```
Comment (tlambert03) [inline napari/layers/image/_image_constants.py]: I'd be more inclined to do this if we were using it for something other than just the GUI... but this also makes it less useful as a "subset"

```python
In [5]: from napari.layers.image._image_constants import Interpolation, InterpolationSubset

In [6]: Interpolation.BILINEAR is InterpolationSubset.BILINEAR
Out[6]: False

In [7]: Interpolation.BILINEAR == InterpolationSubset.BILINEAR
Out[7]: False
```

```suggestion
```

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 2]
```diff
@@ -254,7 +261,9 @@ def _toggle_rendering_parameter_visbility(self):
     def _update_interpolation_combo(self):
         self.interpComboBox.clear()
         interp_enum = (
-            Interpolation3D if self.layer._ndisplay == 3 else Interpolation
+            Interpolation3D
```
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_image_controls.py]: and then just make this:
```python
        interp_2d = ('bicubic', 'bilinear', 'kaiser', 'nearest', 'spline36')
        interp_enum = (
            Interpolation3D.keys() if self.layer._ndisplay == 3 else interp_2d
        )
        self.interpComboBox.addItems(interp_enum)
```

**Review by tlambert03 — COMMENTED** [review]
thanks @alisterburt, the addition of an approved interp works nicely.  I left a couple comments for simplification (I'm not sure we need a new Enum here just for the sake of the combobox, particularly since it doesn't equal it's partner enum)...

as for a test, how about this?

```python
# test_qt_image_base_layer_.py

def test_interpolation_combobox(qtbot):
    """Changing the model attribute should update the view"""
    layer = Image(np.random.rand(8, 8))
    qtctrl = QtImageControls(layer)
    qtbot.addWidget(qtctrl)
    combo = qtctrl.interpComboBox
    opts = {combo.itemText(i) for i in range(combo.count())}
    assert opts == {'bicubic', 'bilinear', 'kaiser', 'nearest', 'spline36'}
    # programattically adding approved interpolation works
    layer.interpolation = 'hanning'
    assert combo.findText('hanning') == 5
```

**Comment by alisterburt** [comment]
@tlambert03 thanks for your help/suggestions - I've proposed two solutions, both work perfectly

1. As you suggested, the view subset is defined as a tuple in qt_image_controls
https://github.com/napari/napari/pull/2552/commits/02601f0c9cdf004d2f0ba7081db82ef2a9821566
2. Alternative where the view subset is defined as a property of the Enum in `_image_constants.py` (idea from [stackoverflow](https://softwareengineering.stackexchange.com/questions/357405/whats-the-most-idiomatic-way-to-make-a-collection-of-enum-enum-in-python)
https://github.com/napari/napari/pull/2552/commits/0188a53bea7384cb779debb7ef488ba350c029ae

I prefer 2 but am very much happy to go with either 🙂 

I added the test in a new `test_qt_image_layer.py` file because interpolation is not part of the base layer

**Comment by tlambert03** [comment]
I like number 2 as well!

File: napari/layers/image/_image_constants.py [hunk 3]
```diff
@@ -28,6 +28,20 @@ class Interpolation(StringEnum):
...
+            self.BILINEAR,
+            self.KAISER,
+            self.NEAREST,
+            self.SPLINE36,
+        )
+
+    @property
+    def in_view_subset(self):
+        return self in self.view_subset
+
```
Comment (tlambert03) [inline napari/layers/image/_image_constants.py]:
```suggestion
    @classmethod
    def view_subset(cls):
        return (
            cls.BICUBIC,
            cls.BILINEAR,
            cls.KAISER,
            cls.NEAREST,
            cls.SPLINE36,
        )
```

it's not quite as useful to access this subset from individual `Interpolation` instances... so maybe lets make it a class method (called as `Interpolation.view_subset()`

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 4]
```diff
@@ -253,10 +260,14 @@ def _toggle_rendering_parameter_visbility(self):
...
     def _update_interpolation_combo(self):
         self.interpComboBox.clear()
-        interp_enum = (
-            Interpolation3D if self.layer._ndisplay == 3 else Interpolation
+        interp_names_2d = (
+            i.name.lower() for i in Interpolation if i.in_view_subset
+        )
+        interp_names_3d = Interpolation3D.keys()
+        interp_names = (
+            interp_names_3d if self.layer._ndisplay == 3 else interp_names_2d
```
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_image_controls.py]:
```suggestion
        interp_names = (
            Interpolation3D.keys()
            if self.layer._ndisplay == 3
            else [i.value for i in Interpolation.view_subset()]
```

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 5]
```diff
@@ -214,10 +214,17 @@ def _on_interpolation_change(self, event):
...
+            return self.interpComboBox.findText(
                 self.layer.interpolation, Qt.MatchFixedString
             )
+
+        with self.layer.events.interpolation.blocker():
+            index = get_index()
+            if index == -1:
+                self.interpComboBox.addItem(self.layer.interpolation)
+                index = get_index()
             self.interpComboBox.setCurrentIndex(index)
```
Comment (tlambert03) [inline napari/_qt/layer_controls/qt_image_controls.py]:
```suggestion
        interp_string = event.value.value

        with self.layer.events.interpolation.blocker():
            if self.interpComboBox.findText(interp_string) == -1:
                self.interpComboBox.addItem(interp_string)
            self.interpComboBox.setCurrentText(interp_string)
```

and then, in `layers/image/image.py` ...  around line 438
```python
# change this
self.events.interpolation()
# to this
self.events.interpolation(value=self._interpolation[self._ndisplay])

**Review by sofroniewn — APPROVED** [review]
LGTM! `view_subset` seems like a reasonable approach to take here.

File: napari/_qt/layer_controls/qt_image_controls.py [hunk 6]
```diff
@@ -214,11 +214,12 @@ def _on_interpolation_change(self, event):
...
             The napari event that triggered this method.
         """
+        interp_string = event.value.value
+
         with self.layer.events.interpolation.blocker():
-            index = self.interpComboBox.findText(
-                self.layer.interpolation, Qt.MatchFixedString
-            )
-            self.interpComboBox.setCurrentIndex(index)
+            if self.interpComboBox.findText(interp_string) == -1:
```
Comment (alisterburt) [inline napari/_qt/layer_controls/qt_image_controls.py:220]: so much cleaner than how I did it! thanks for going over this @tlambert03

**Comment by ambientnuance** [comment]
If you don't mind me popping in here, I'd like to table an alternative approach that came to mind when reading the motivating issue for this PR. It may be applicable/better suited to other scenarios.

Simplifying the drop-list is a good idea, but it doesn't necessarily require removing GUI access to the full suite. A '...' option could be appended to the menu - clicking this would present any remaining options; clicking again would collapse back to the initial subset. This approach is motivated by a [quote][1] from Alan Kay and can be taken further...

Interpolation methods that are parameterised could allow for tuning within the GUI. But, in a similar vein as above, this functionality wouldn't be presented to the user by default. A button ('...' or an icon of analogue dial) placed adjacent to the drop-list would expand/collapse the available parameters. If the selected method is non-parameterised, this button would be faded out.

On the informational side of things, a similarly behaving button (the classic 'information i') could toggle the display of explanatory notes and/or links. The content could be sourced from the corresponding docstring, or from some common resource shared by the docstring.

---

For each idea, feel free to take it or leave it - no worries either way.

If any take your fancy, let's talk implementation. I haven't worked with PyQt/Qt before - do you think these kind of features would be approachable for me to implement? At least the extensible drop-list to begin with.

---

A couple of misc. thoughts that personally motivate this approach:

- Features that are hidden or inaccessible for a user are less likely to be tested 'in the wild'.

- I've found that the topic of interpolation (and the methods involved) is often laced with subtlety. Allowing a user to surface some of the details involved could help them get a better grasp of how things work/which methods are appropriate for which situation.

[1]: _ "Simple things should be simple, complex things should be possible"

**Pull request merged**